okio/src/main/java/okio/RealBufferedSource.java

```diff
@@ -442,4 +471,54 @@ public boolean rangeEquals(long offset, ByteString bytes, int bytesOffset, int b
   @Override public String toString() {
     return "buffer(" + source + ")";
   }
+
+  final class BomAwareReader extends Reader {
+    final Charset suppliedCharset;
+    final ByteOrderMark[] byteOrderMarks;
+    private Reader delegate;
+
+    BomAwareReader(Charset charset, ByteOrderMark... byteOrderMarks) {
+      this.suppliedCharset = charset;
+      this.byteOrderMarks = byteOrderMarks;
+    }
+
+    @Override public int read(char[] data, int offset, int charCount) throws IOException {
+      if (delegate == null) {
+        if (closed) throw new IOException("closed");
+
+        // Make sure buffer holds all bytes of the BOM.
+        if (buffer.size() < ByteOrderMark.MAX_BOM_LENGTH) {
```

Comment:
Instead you need to peek a single byte, and make a decision about whether to read more characters or not.